validator_client/slashing_protection/src/slashing_protection.rs

```diff
@@ -0,0 +1,378 @@
+use crate::attester_slashings::{check_for_attester_slashing, SignedAttestation};
+use crate::enums::{NotSafe, Safe, ValidityReason};
+use crate::proposer_slashings::{check_for_proposer_slashing, SignedBlock};
+use fs2::FileExt;
+use ssz::{Decode, Encode};
+use std::convert::TryFrom;
+use std::fs::File;
+use std::io::{ErrorKind, Read, Result as IOResult, Write};
+use std::os::unix::fs::PermissionsExt;
+use std::path::{Path, PathBuf};
+use types::{AttestationData, BeaconBlockHeader};
+
+/// Trait used to know if type T can be checked for slashing safety
+pub trait SafeFromSlashing<T> {
+    type U;
+
+    /// Verifies that the incoming_data is not slashable and returns
+    /// the index at which it should get inserted in the history.
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &Self::U,
+        data_history: &[T],
+    ) -> Result<Safe, NotSafe>;
+
+    /// Checks if incoming_data is free from slashings, and if so updates the in-memory and writes it to the history file.
+    /// If the error returned is an IOError, do not sign nor broadcast the attestation.
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe>;
+}
+
+impl SafeFromSlashing<SignedAttestation> for HistoryInfo<SignedAttestation> {
+    type U = AttestationData;
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &AttestationData,
+        data_history: &[SignedAttestation],
+    ) -> Result<Safe, NotSafe> {
+        check_for_attester_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
```

Comment:
I would've liked to have this method be a generic one, but I didn't manage to do so